pallets/fusion/src/lib.rs

```diff
@@ -0,0 +1,1989 @@
+	#[pallet::storage]
+	#[pallet::getter(fn claimed_rewards)]
+	pub type ClaimedRewards<T: Config> = StorageNMap<
+		_,
+		(
+			NMapKey<Blake2_128Concat, EvmAddress>,
+			NMapKey<Blake2_128Concat, PoolId>,
+			NMapKey<Blake2_128Concat, EraIndex>,
+		),
+		BalanceOf<T>,
+		OptionQuery,
+	>;
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// Event triggered when the funding account is filled with new funds
+		FundsAccountFilled {
+			sender: T::AccountId,
+			amount: BalanceOf<T>,
+		},
+		/// Event triggered when a new currency is created
+		CurrencyCreated {
+			currency_id: CurrencyId,
+			name: BoundedVec<u8, T::MaxCurrencyName>,
+			nb_decimals: u8,
+			max_amount: FusionCurrencyBalance,
+			min_amount: FusionCurrencyBalance,
+			initial_conversion_rate: BalanceOf<T>,
+		},
+		/// Event triggered when a currency's properties are updated
+		CurrencySet {
+			currency_id: CurrencyId,
+			name: Option<BoundedVec<u8, T::MaxCurrencyName>>,
+			max_amount: Option<FusionCurrencyBalance>,
+			min_amount: Option<FusionCurrencyBalance>,
+		},
+		/// Event triggered when a currency is deleted
+		CurrencyDeleted { currency_id: CurrencyId },
+		/// Event triggered when a conversion rate is set for a currency
+		CurrencyConversionRateSet {
+			currency_id: CurrencyId,
+			conversion_rate: BalanceOf<T>,
+		},
+		/// Event triggered when a currency is deposited into the system
+		CurrencyDeposited {
+			evm_address: EvmAddress,
+			currency_id: CurrencyId,
+			amount: FusionCurrencyBalance,
+		},
+		/// Event triggered when a user unbonds currency from a pool
+		CurrencyUnbonded {
+			evm_address: EvmAddress,
+			pool_id: PoolId,
+			currency_id: CurrencyId,
+			unbonded_amount: FusionCurrencyBalance,
+			points: Points,
+			era: EraIndex,
+		},
+		/// Event triggered when a user withdraws unbonded currency
+		CurrencyWithdrawn {
+			evm_address: EvmAddress,
+			pool_id: PoolId,
+			currency_id: CurrencyId,
+			amount: FusionCurrencyBalance,
+		},
+		/// Event triggered when the controller address for a user is changed
+		ControllerAddressSet {
+			evm_address: EvmAddress,
+			new_controller_address: Option<T::AccountId>,
+		},
+		/// Event triggered when the compounding value is changed for a pool member
+		CompoundingSet {
+			evm_address: EvmAddress,
+			pool_id: PoolId,
+			compound: bool,
+		},
+		/// Event triggered when a new Fusion pool is created
+		PoolCreated {
+			pool_id: PoolId,
+			currency_id: CurrencyId,
+			apy: Perbill,
+			state: FusionPoolState,
+			nominator: Option<T::AccountId>,
+			reward_account: Option<T::AccountId>,
+		},
+		/// Event triggered when a Fusion pool's properties are updated
+		PoolSet {
+			pool_id: PoolId,
+			apy: Option<Perbill>,
+			state: Option<FusionPoolState>,
+			nominator: Option<Option<T::AccountId>>,
+		},
+		/// Event triggered when a user joins a pool
+		PoolJoined {
+			evm_address: EvmAddress,
+			pool_id: PoolId,
```

Comment:
For sake of consistency, it could be useful to put `pool_id` as the first member of all events.